### PR TITLE
Avoid copy in `chol_lower`

### LINKED ISSUE #143: `chol_lower` does not avoid allocations
https://github.com/JuliaStats/PDMats.jl/blob/dfcf96c90ade0ad6052855a3e4b3488b3dbe002a/src/chol.jl#L2 always has a superfluous copy, even though the point of defining `chol_lower` in this package is to avoid the extra allocation (as mentioned in the comment a few lines further down), to work around the issue that https://github.com/JuliaLang/julia/pull/42920 tries to address upstream.

MWE:
```julia
julia> using LinearAlgebra, PDMats

julia> A = rand(100, 100); C = A'A;

julia> PDMats.chol_lower(C); PDMats.chol_lower(cholesky(C)); cholesky(C).U';

julia> @allocated PDMats.chol_lower(C)
160240

julia> @allocated PDMats.chol_lower(cholesky(C))
80192

julia> @allocated cholesky(C).U'
80192
```

### PR DESCRIPTION
**Pull request by st--**
Fixes #143

### CODE REVIEW

**Comment by st--** [comment]
Side note: `chol_lower(::CholTypeSparse)` *seems* to be working fine, but I'm not sure I understand enough about the implementation details of SuiteSparse to guarantee this.

File: src/chol.jl [hunk 1]
```diff
@@ -1,9 +1,9 @@
-CholType{T,S<:AbstractMatrix} = Cholesky{T,S}
```
Comment (st--) [inline src/chol.jl:1]: This seems to be entirely redundant, so I changed it to use Cholesky directly instead.

**Review by andreasnoack — COMMENTED** [review]
I think you can just make this `chol_lower(a::Matrix) = cholesky(Hermitian(a, :L)).L`.

I agree that `CholType` is redundant but please make that change in a separate PR.

File: src/chol.jl (hunk 1 above)
Comment (st--) [reply src/chol.jl:1]: Moved into #145

**Comment by st--** [comment]
> I agree that `CholType` is redundant but please make that change in a separate PR.

Done, see #145

**Comment by st--** [comment]
> I think you can just make this `chol_lower(a::Matrix) = cholesky(Hermitian(a, :L)).L`.

Writing it in terms of `chol_lower(cholesky(a))` seems cleaner to me, and it should compile to the same code; what benefits do you see in writing it as this ?

**Comment by andreasnoack** [comment]
> and it should compile to the same code

No. The default is upper
```julia
julia> cholesky(A).uplo
'U': ASCII/Unicode U+0055 (category Lu: Letter, uppercase)

julia> cholesky(Hermitian(A, :L)).uplo
'L': ASCII/Unicode U+004C (category Lu: Letter, uppercase)
```

**Comment by andreasnoack** [comment]
It's just an artifact of the not wrapping it in a function
```julia
julia> chol_lower_an(A::Matrix) = cholesky(Hermitian(A, :L)).L;

julia> @allocated chol_lower_an(A);

julia> @allocated chol_lower_an(A)
80144
```
It's also best to avoid wrapper types when possible.

**Comment by st--** [comment]
> It's just an artifact of the not wrapping it in a function

Ah, thanks for pointing that out.

> It's also best to avoid wrapper types when possible.

Could you expand on what you mean ?

**Comment by andreasnoack** [comment]
> Could you expand on what you mean ?

Sure. The more types layers you have the less likely it is that a specific method has been written for it and you risk hitting a less optimized fallback. The compiler can also sometimes completely remove these wrapper types but I believe that the more layers you have, the harder it is for the compiler to do so.

**Comment by st--** [comment]
Ok, so do I understand correctly, you would prefer
```diff
-chol_lower(a::Matrix) = chol_lower(cholesky(a))
+chol_lower(a::Matrix) = cholesky(Hermitian(a, :L)).L
```
because the former returns a `LowerTriangular{Adjoint{Matrix}}`, whereas the latter returns a `LowerTriangular{Matrix}` directly?

**Comment by andreasnoack** [comment]
That is correct

**Comment by st--** [comment]
Thanks, that makes sense. I've changed it accordingly:)

**Review by andreasnoack — APPROVED** [review]

**Comment by st--** [comment]
Downside of merging it in with `Hermitian()` is that this breaks AD :S

**Comment by andreasnoack** [comment]
Oh no. Do you have an example?

**Comment by st--** [comment]
Yes, it turns out to be a Zygote issue, I've made a PR to fix it here: https://github.com/FluxML/Zygote.jl/pull/1114

**Comment by st--** [comment]
@andreasnoack looks like the Zygote issue might be a bit more complex to resolve; how about (at least temporarily) changing it from `Hermitian` to `Symmetric`?

**Comment by andreasnoack** [comment]
Given that we have this restriction https://github.com/JuliaStats/PDMats.jl/blob/635a957abc71aa9b63d276f96a185db1bed9de74/src/PDMats.jl#L36 it should be fine

**Pull request merged**